src/pages/playground/Playground.tsx

```diff
@@ -732,85 +689,62 @@ const Playground: React.FC<PlaygroundProps> = ({ workspaceLocation = 'playground
   ]);
 
   const playgroundIntroductionTab: SideContentTab = React.useMemo(
-    () => ({
-      label: 'Introduction',
-      iconName: IconNames.HOME,
-      body: (
-        <Markdown
-          content={generateSourceIntroduction(
-            props.playgroundSourceChapter,
-            props.playgroundSourceVariant
-          )}
-          openLinksInNewWindow={true}
-        />
+    () =>
+      makeIntroductionTabFrom(
+        generateSourceIntroduction(props.playgroundSourceChapter, props.playgroundSourceVariant)
       ),
-      id: SideContentType.introduction
-    }),
     [props.playgroundSourceChapter, props.playgroundSourceVariant]
   );
 
   const tabs = React.useMemo(() => {
     const tabs: SideContentTab[] = [playgroundIntroductionTab];
 
-    // For HTML Chapter, HTML Display tab is added only after code is run
-    if (props.playgroundSourceChapter === Chapter.HTML) {
-      if (props.output.length > 0 && props.output[0].type === 'result') {
-        tabs.push({
-          label: 'HTML Display',
-          iconName: IconNames.MODAL,
-          body: (
-            <SideContentHtmlDisplay
-              content={(props.output[0] as ResultOutput).value}
-              handleAddHtmlConsoleError={errorMsg =>
-                dispatch(addHtmlConsoleError(errorMsg, workspaceLocation))
-              }
-            />
-          ),
-          id: SideContentType.htmlDisplay
-        });
-      }
-      return tabs;
-    }
-
-    // (TEMP) Remove tabs for fullJS until support is integrated
-    if (
-      props.playgroundSourceChapter === Chapter.FULL_JS ||
-      props.playgroundSourceChapter === Chapter.FULL_TS
-    ) {
-      return [...tabs, dataVisualizerTab];
-    }
+    const currentLang = props.playgroundSourceChapter;
+    const currentVariant = props.playgroundSourceVariant;
+
+    switch (true) {
+      case !isSourceLanguage(currentLang):
+        // For now, disable other tabs when not running Source
+        return tabs;
+      case currentLang === Chapter.HTML:
+        // For HTML Chapter, HTML Display tab is added only after code is run
+        if (props.output.length > 0 && props.output[0].type === 'result') {
+          tabs.push(
+            makeHtmlDisplayTabFrom(props.output[0] as ResultOutput, errorMsg =>
+              dispatch(addHtmlConsoleError(errorMsg, workspaceLocation))
+            )
+          );
+        }
+        return tabs;
+      case currentLang === Chapter.FULL_JS:
+      case currentLang === Chapter.FULL_TS:
+        // (TEMP) Remove tabs for fullJS until support is integrated
+        return [...tabs, dataVisualizerTab];
+      case !usingRemoteExecution:
+        // Don't show the following when using remote execution
+
+        // Enable Data Visualizer for Source Chapter 2 and above
+        const shouldShowDataVisualizer = currentLang >= Chapter.SOURCE_2;
+        if (shouldShowDataVisualizer) {
+          tabs.push(dataVisualizerTab);
+        }
```

Comment:
> I know converting the if-else statements into switch-case blocks somewhat makes it less readable, but this is a temporary change as once the logic for rendering is abstracted out, I promise it would be much cleaner.

I won't block on this since it's a temporary change, but wouldn't writing this using chained if statements be easier to read without having to perform huge restructuring the code here? Using a switch statement is a bit weird since conventionally, the expression to be matched is the one whose values changes, and not the case clauses.

Reply:
Actually yea, on second thought, I might revert this in the future PR. Thanks!

Reply:
Changed in 5fe986b74d4cebdadf9a930481e9c643546d558d.